### PR TITLE
SONARJAVA-1876 Rule S1820: Classes should not have too many fields

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/ClassFieldCountCheck.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      .filter(member -> member.is(Tree.Kind.VARIABLE) && shouldBeCounted((VariableTree) member))
+      .count();
+    if (fieldCount > threshold) {
+      String message = String.format("Refactor this class so it has no more than %d %sfields, rather than the %d it currently has.", threshold,
+        countNonPublicFields ? "" : "public ", fieldCount);
+      reportIssue(treeToReport(classTree), message);
+    }
+  }
+
+  private Tree treeToReport(ClassTree classTree) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/ClassFieldCountCheck.java]: MAJOR Make "treeToReport" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/ClassFieldCountCheck.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    if (fieldCount > threshold) {
+      String message = String.format("Refactor this class so it has no more than %d %sfields, rather than the %d it currently has.", threshold,
+        countNonPublicFields ? "" : "public ", fieldCount);
+      reportIssue(treeToReport(classTree), message);
+    }
+  }
+
+  private static Tree treeToReport(ClassTree classTree) {
+    //simpleName can be null for anonymous classes
+    return MoreObjects.firstNonNull(classTree.simpleName(), classTree.parent());
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ClassFieldCountCheck.java]: You should use `org.sonar.java.checks.helpers.ExpressionsHelper.reportOnClassTree(ClassTree)` rather that getting the `classTree.parent()`. For anonymous classes, you will end up with the whole class being highlighted (big read area), which is not what we want.

File: java-checks/src/test/files/checks/ClassFieldCountCheck/UnusualTrees.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+interface TooManyFieldsInterface { // Noncompliant [[sc=11;ec=33]] {{Refactor this class so it has no more than 2 fields, rather than the 3 it currently has.}}
+   int field1;
+   int field2;
+   int field3;
+}
+
+class TooManyFieldsAnonymous {
+  {
+    new java.util.ArrayList<Integer>() { // Noncompliant [[sc=5]] {{Refactor this class so it has no more than 2 fields, rather than the 3 it currently has.}}
```
Comment (m-g-sonar) [inline java-checks/src/test/files/checks/ClassFieldCountCheck/UnusualTrees.java]: You should have checked for for the `end column`. Specifying `end line` with `el` can help.

Usually, we try to limit as much as possible large area being highlighted. In that specific case, highlighting everything, starting from the new keyword and including the whole body is too much. For anonymous classes, to only highlight the type tree is enough.

File: java-checks/src/test/files/checks/ClassFieldCountCheck/UnusualTrees.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+enum TooManyFieldsEnum { // Noncompliant [[sc=6;ec=23]] {{Refactor this class so it has no more than 2 fields, rather than the 3 it currently has.}}
+  ENUM_CONSTANT_TREE { // Noncompliant [[sc=3;ec=4]] {{Refactor this class so it has no more than 2 fields, rather than the 3 it currently has.}}
```
Comment (m-g-sonar) [inline java-checks/src/test/files/checks/ClassFieldCountCheck/UnusualTrees.java]: Same remark than for the anonymous class. we should only highlight the name. When different lines are involved, please use the `endLine` specifier (`[[sc=3;el=+4;ec=4]]`).

File: java-checks/src/main/java/org/sonar/java/checks/ClassFieldCountCheck.java (hunk 2 above)
Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/ClassFieldCountCheck.java]: Done

**Pull request merged**